### PR TITLE
mgr/ssh: packaged-ceph-daemon mode; ssh key mgmt

### PR DESCRIPTION
**Pull request by liewegas**
- Add mgr/ssh commands to manage the ssh key
- ceph-daemon uses mgr/ssh to generate key
- add mode to use an installed ceph-daemon, which cephdaemon user + sudo

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard backend`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Comment by sebastian-philipp** [comment]
@smithfarm , any thoughts on this one?

File: src/pybind/mgr/ssh/module.py [hunk 1]
```diff
@@ -35,6 +36,29 @@
...
+
+    return CLICommand(prefix, cmd_args, desc, perm)(wrapper)
+
+def _cli_command(perm):
+    def inner_cli_command(prefix, cmd_args="", desc=""):
+        return lambda func: handle_exception(prefix, cmd_args, desc, perm, func)
+    return inner_cli_command
+
+_read_cli = _cli_command('r')
+_write_cli = _cli_command('rw')
```
Comment (sebastian-philipp) [inline src/pybind/mgr/ssh/module.py]: (feeling kinda honored for making us of it, but)

Copied from `orchestrator_cli`.  I'd be :+1: to have this available in the `orchestrator.py` interface.

File: src/pybind/mgr/ssh/module.py [hunk 2]
```diff
@@ -350,16 +369,63 @@ def _clear_ssh_config(self):
...
+            path = tmp_dir.name + '/key'
+            subprocess.call([
+                '/usr/bin/ssh-keygen',
+                '-C', 'ceph-%s' % self._cluster_fsid,
+                '-N', '',
+                '-f', path
+            ])
+            with open(path, 'r') as f:
+                secret = f.read()
+            with open(path + '.pub', 'r') as f:
```
Comment (sebastian-philipp) [inline src/pybind/mgr/ssh/module.py]: if this raises, `os.unlink(path)` is never executed. 

```python
try:
    with open(path + '.pub', 'r') as f:
        pub = f.read()
    os.unlink(path + '.pub')
finally:
    os.unlink(path)
```
?

**Comment by sebastian-philipp** [comment]
cc @swiftgist

**Comment by smithfarm** [comment]
@sebastian-philipp Someone will need to add RPM scriptlets, I guess... This PR appears to be Debian-only.

**Comment by sebastian-philipp** [comment]
failed: https://jenkins.ceph.com/job/ceph-dev-new-build/ARCH=x86_64,AVAILABLE_ARCH=x86_64,AVAILABLE_DIST=leap15,DIST=leap15,MACHINE_SIZE=huge/33086//consoleFull

File: ceph.spec.in [hunk 3]
```diff
@@ -1492,8 +1493,27 @@ if [ $1 -ge 1 ] ; then
...
+%if 0%{?rhel} || 0%{?fedora}
+/usr/sbin/useradd cephdaemon -r -s /bin/bash -c "Ceph-daemon user for mgr/ssh" -m /home/cephdaemon
+%endif
+%if 0%{?suse_version}
+if ! getent passwd cephdaemon >/dev/null ; then
+    useradd cephdaemon -r -s /bin/bash -m /home/cephdaemon 2>/dev/null || :
+fi
+usermod -c "Ceph-daemon user for mgr/ssh" \
+        -s /bin/bash \
+        cephdaemon
```
Comment (tchaikov) [inline ceph.spec.in]:
```
usermod: user 'cephdaemon' does not exist
```

**Comment by liewegas** [comment]
Consider this RFC--I wasn't able to test any of the packaging changes yet

**Comment by liewegas** [comment]
Retested and the packages appear to be behaving!

File: src/pybind/mgr/ssh/module.py [hunk 4]
```diff
@@ -411,19 +461,23 @@ def _run_ceph_daemon(self, host, entity, command, args,
...
+                script += self._ceph_daemon
+                out, err, code = remoto.process.check(
+                    conn,
+                    ['/usr/bin/python', '-u'],
+                    stdin=script.encode('utf-8'))
+            elif self.mode == 'ceph-daemon-package':
+                out, err, code = remoto.process.check(
+                    conn,
+                    ['sudo', '/usr/bin/ceph-daemon'] + final_args,
+                    stdin=stdin)
```
Comment (sebastian-philipp) [inline src/pybind/mgr/ssh/module.py]: Yep, that was my thought too.

**Comment by liewegas** [comment]
I have a follow-on trello card to add ceph2.py support for this mode, but i don't want to hold up either of these PRs for that.

**Pull request merged**